{% extends "layout.html" %}
{% set title = _('ProPhyle: DNA sequence classification') %}
{% block body %}

  <h1>Welcome to ProPhyle</h1>

  <p>
    ProPhyle brings metagenomic classification <b>from clusters to laptops</b>. This is possible thanks to a novel indexing strategy,
    based on the bottom-up propagation of k-mers in the phylogenetic/taxonomic tree, assembling contigs at each node and matching using a full-text search.
  </p>

  <p>Compared to other state-of-the-art classifiers, ProPhyle provides several unique features:

  <img src="{{ pathto('_static/prophyle_overview.jpg', 1) }}"  style="float:right; width:250px" />

  <ul>
    <li><b>Low memory requirements.</b> Compared to Kraken, ProPhyle has 7x smaller memory footprint for index construction and 5x smaller footprint for querying, while providing a more expressive index.</li>
    <li><b>Flexibility.</b> ProPhyle is easy to use with any user-provided phylogenetic trees and reference genomics sequences (e.g., reads or assemblies). It can classify short reads, long reads, or even assembled contigs.</li>
    <li><b>Standard bioinformatics formats.</b> Newick/NHX is used for representing phylogenetic trees and SAM/BAM for reporting assignments.</li>
    <li><b>Lossless k-mer indexing.</b> ProPhyle stores a list of all genomes containing a k-mer. The classification is, therefore, accurate even with trees containing similar genomes (e.g, phylogenetic trees for a single species).</li>
    <li><b>Reproducibility.</b> ProPhyle is fully deterministic, with a mathematically well-defined behavior. Databases are versioned and distributed via Zenodo.</li>
  </ul>

  <h2 style="margin-bottom: 0">Documentation</h2>

  <table class="contentstable">
    <tr>
      <td style="width:35%">
        <p class="biglink"><a class="biglink" href="{{ pathto("example") }}">Quick example</a><br/>
           <span class="linkdescr">5 minutes example</span></p>
      </td><td>
        <p class="biglink"><a class="biglink" href="{{ pathto("contents") }}">Contents</a><br/>
           <span class="linkdescr">for a complete overview</span></p>
      </td>
    </tr><tr>
      <td>
        {%- if hasdoc('search') %}<p class="biglink"><a class="biglink" href="{{ pathto("search") }}">Search page</a><br/>
           <span class="linkdescr">search the documentation</span></p>{%- endif %}
      </td><td>
        <p class="biglink"><a class="biglink" href="https://github.com/prophyle/prophyle/releases">
          Releases
        </a><br/>
           <span class="linkdescr">release history</span></p>
      </td>
    </tr>
  </table>


  <h2 style="margin-bottom: 0">Auxiliary tools</h2>

  <table class="contentstable">
    <tr>
      <td style="width:35%">
        <p class="biglink"><a class="biglink" href="http://github.com/prophyle/prophex">ProphEx</a><br/>
           <span class="linkdescr">FM-index for rapid k-mer matching using simplitigs</span>
         </p>
      </td>
      <td>
        <p class="biglink"><a class="biglink" href="http://github.com/prophyle/prophasm">ProphAsm</a><br/>
           <span class="linkdescr">representing k-mer sets via simplitigs</span>
         </p>
      </td>
    </tr>
  </table>


  <h2>Cite</h2>

  <div class="cit">
  <p>
	[1] K. Břinda, L. Lima, S. Pignotti, N. Quinones-Olvera, K. Salikhov,
	R. Chikhi, G. Kucherov, Z. Iqbal, and M. Baym,
	<b>Efficient and robust search of microbial genomes via phylogenetic compression,</b>
    <a href="https://doi.org/10.1101/2023.04.15.536996">https://doi.org/10.1101/2023.04.15.536996</a>.
  </p>
  <p>
    [2] Břinda K, Salikhov K, Pignotti S, Kucherov G.
    <b>ProPhyle 0.3.1.0</b>, Zenodo, 2017.
    <a href="http://doi.org/10.5281/zenodo.1045429">https://doi.org/10.5281/zenodo.1045429</a>.
  </p>
  <p>
    [3] Břinda K, Salikhov K, Pignotti S, Kucherov G.
    <b>ProPhyle: a phylogeny-based metagenomic classifier using the Burrows-Wheeler Transform.</b>
    Poster at HiTSeq 2017.
    <a href="http://doi.org/10.5281/zenodo.1045427">https://doi.org/10.5281/zenodo.1045427</a>
  </p>
  <p>
    [4] Břinda K.
    <b>Novel computational techniques for mapping and classifying Next-Generation Sequencing data.</b>
    PhD Thesis, Université Paris-Est, 2016.
    <a href="http://doi.org/10.5281/zenodo.1045317">https://doi.org/10.5281/zenodo.1045317</a>
  </p>
  <p>
    [5] Salikhov K.
    <b>Efficient algorithms and data structures for indexing DNA sequence data.</b>
    PhD Thesis, Université Paris-Est, 2017.
  </p>
</div>

<p style="font-size:90%">
    [1] is the main reference for the entire ProPhyle package. [2] contains a summary of the ProPhyle algorithm, [3] provides a thorough description (see Chapter 12), and [4] explains details of the BWT-indexing technique.
</p>


  <h2>Authors</h2>

  <ul>
  <li><a href="http://brinda.eu">Karel Břinda</a> &lt;<a href="mailto:karel.brinda@inria.fr">karel.brinda@inria.fr</a>&gt;</li>
  <li>Kamil Salikhov &lt;<a href="mailto:kamil.salikhov@univ-mlv.fr">kamil.salikhov@univ-mlv.fr</a>&gt;</li>
  <li>Simone Pignotti &lt;<a href="mailto:pignottisimone@gmail.com">pignottisimone@gmail.com</a>&gt;</li>
  <li><a href="http://igm.univ-mlv.fr/~koutcher/">Gregory Kucherov</a> &lt;<a href="mailto:gregory.kucherov@univ-mlv.fr">gregory.kucherov@univ-mlv.fr</a>&gt;</li>
  </ul>

{% endblock %}
